rskj-core/src/main/java/co/rsk/peg/federation/FederationStorageProviderImpl.java

```diff
@@ -0,0 +1,374 @@
+        saveOldFederationBtcUTXOs();
+        saveNewFederation(activations);
+        saveOldFederation(activations);
+        //savePendingFederation(activations);
+        saveFederationElection();
+        saveActiveFederationCreationBlockHeight(activations);
+        saveNextFederationCreationBlockHeight(activations);
+        saveLastRetiredFederationP2SHScript(activations);
+    }
+
+    private void saveNewFederationBtcUTXOs(NetworkParameters networkParameters, ActivationConfig.ForBlock activations) {
+        if (newFederationBtcUTXOs == null) {
+            return;
+        }
+
+        DataWord key = getStorageKeyForNewFederationBtcUtxos(networkParameters, activations);
+        bridgeStorageAccessor.safeSaveToRepository(key, newFederationBtcUTXOs, BridgeSerializationUtils::serializeUTXOList);
+    }
+    private void saveOldFederationBtcUTXOs() {
+        if (oldFederationBtcUTXOs == null) {
+            return;
+        }
+
+        bridgeStorageAccessor.safeSaveToRepository(OLD_FEDERATION_BTC_UTXOS_KEY.getKey(), oldFederationBtcUTXOs, BridgeSerializationUtils::serializeUTXOList);
+    }
+
+    private void saveNewFederation(ActivationConfig.ForBlock activations) {
+        if (newFederation == null) {
+            return;
+        }
+
+        StorageAccessor.RepositorySerializer<Federation> serializer = BridgeSerializationUtils::serializeFederationOnlyBtcKeys;
+
+        if (activations.isActive(RSKIP123)) {
+            saveStorageVersion(
+                NEW_FEDERATION_FORMAT_VERSION.getKey(),
+                newFederation.getFormatVersion()
+            );
+            serializer = BridgeSerializationUtils::serializeFederation;
+        }
+
+        bridgeStorageAccessor.safeSaveToRepository(NEW_FEDERATION_KEY.getKey(), newFederation, serializer);
+    }
+
+    private void saveOldFederation(ActivationConfig.ForBlock activations) {
+        if (!shouldSaveOldFederation) {
+            return;
+        }
+
+        StorageAccessor.RepositorySerializer<Federation> serializer = BridgeSerializationUtils::serializeFederationOnlyBtcKeys;
+
+        if (activations.isActive(RSKIP123)) {
+            if (oldFederation != null) {
+                saveStorageVersion(
+                    OLD_FEDERATION_FORMAT_VERSION.getKey(),
+                    oldFederation.getFormatVersion()
+                );
+            } else {
+                // assume it is a standard federation to keep backwards compatibility
+                saveStorageVersion(
+                    OLD_FEDERATION_FORMAT_VERSION.getKey(),
+                    STANDARD_MULTISIG_FEDERATION.getFormatVersion()
+                );
+            }
+            serializer = BridgeSerializationUtils::serializeFederation;
+        }
+
+        bridgeStorageAccessor.safeSaveToRepository(OLD_FEDERATION_KEY.getKey(), oldFederation, serializer);
+    }
+
+    // TODO
+/*    private void savePendingFederation() throws IOException {
```

Comment:
Is this what we have pending to discuss?

Reply:
yes